### PR TITLE
Fix `PropertyListHelper::_get_property` returning a valid value even if an index is outside the array valid indices

### LINKED ISSUE #91504: error when changing TileMap to TileMapLayer
### Tested versions

4.3 dev 6

### System information

windows 10 vulkan foward+

### Issue description

Everything disapears 
when changing from tileMap to tileMapLayer

### Steps to reproduce

- add tileMap node
- place tiles on Fayehs
- changetype tileMap node to tileMapLayers

```
scene/2d/tile_map.cpp:415 - Index p_layer = 1 is out of bounds ((int)layers.size() = 1).
scene/2d/tile_map.cpp:139 - Index p_layer = 1 is out of bounds ((int)layers.size() = 1).
scene/2d/tile_map.cpp:415 - Index p_layer = 2 is out of bounds ((int)layers.size() = 1).
scene/2d/tile_map.cpp:139 - Index p_layer = 2 is out of bounds ((int)layers.size() = 1).
```

### Minimal reproduction project (MRP)

[tilelayerconvert.zip](https://github.com/godotengine/godot/files/15197077/tilelayerconvert.zip)

### PR DESCRIPTION
**Pull request by groud**
Closes #91504

### CODE REVIEW

**Comment by KoBeWi** [comment]
Replaces StringName with MethodBind:

[tilebinds.diff.txt](https://github.com/godotengine/godot/files/15263279/tilebinds.diff.txt)

**Comment by groud** [comment]
Modified with your suggestions. I added you as a co-author.

File: scene/2d/tile_map.cpp [hunk 1]
```diff
@@ -412,6 +412,9 @@ void TileMap::set_layer_name(int p_layer, String p_name) {
 }
 
 String TileMap::get_layer_name(int p_layer) const {
+	if (p_layer >= (int)layers.size()) {
+		print_line("Error");
+	}
```
Comment (KoBeWi) [inline scene/2d/tile_map.cpp:417]:
```suggestion
```
Leftover?

File: editor/scene_tree_dock.cpp [hunk 2]
```diff
@@ -2865,21 +2865,24 @@ void SceneTreeDock::replace_node(Node *p_node, Node *p_by_node) {
...
 
 		for (const PropertyInfo &E : pinfo) {
 			if (!(E.usage & PROPERTY_USAGE_STORAGE)) {
 				continue;
 			}
 
-			if (default_oldnode->get(E.name) != n->get(E.name)) {
-				newnode->set(E.name, n->get(E.name));
+			bool valid;
+			Variant default_val = default_oldnode->get(E.name, &valid);
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2883]:
```suggestion
			const Variant &default_val = default_oldnode->get(E.name, &valid);
```

File: scene/property_list_helper.cpp [hunk 3]
```diff
@@ -83,7 +91,13 @@ bool PropertyListHelper::is_initialized() const {
 }
 
 void PropertyListHelper::setup_for_instance(const PropertyListHelper &p_base, Object *p_object) {
+	ERR_FAIL_COND(p_base.prefix.is_empty());
+	ERR_FAIL_NULL(p_base.array_length_getter);
+	ERR_FAIL_COND(p_base.property_list.is_empty());
+	ERR_FAIL_NULL(p_object);
```
Comment (KoBeWi) [inline scene/property_list_helper.cpp:97]: These could be DEV_ASSERTS tbh. No reason to check it for every single instance.

**Review by KoBeWi — APPROVED** [review]

File: scene/property_list_helper.h [hunk 4]
```diff
@@ -43,15 +43,22 @@ class PropertyListHelper {
...
 	void _call_setter(const MethodBind *p_setter, int p_index, const Variant &p_value) const;
 	Variant _call_getter(const Property *p_property, int p_index) const;
+	int _call_array_length_getter() const;
 
 public:
 	void set_prefix(const String &p_prefix);
+	template <typename G>
+	void set_array_length_getter(G p_array_length_getter) {
+		array_length_getter = create_method_bind(p_array_length_getter);
+	};
```
Comment (AThousandShips) [inline scene/property_list_helper.h:60]:
```suggestion
	}
```

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**